Blog App - processing data in models

Elhambasir opened the pull request:
### What is done in this PR?
Set up the models.

- [x] Created model classes for all entities.
- [x] Set up associations between models.
- [x] Created two users.
- [x] Created posts.
- [x] Created comments.
- [x] Created custom methods
  
  - Users
        A method that returns the 3 most recent posts for a given user.
  - Posts
        A method that updates the posts counter for a user.
        A method which returns the 5 most recent comments for a given post.
  - Comments
        A method that updates the comments counter for a post.
  - Likes
        A method that updates the likes counter for a post.

All methods are checked.

Check project [requirements](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/Processing_data_model_project_v1_1.md)

### Note: This is my last review request. Please call me on slack if you face any issue during running the project.

Thank you for reviewing.

Elhambasir committed:
Create user.rb

Elhambasir committed:
Create post.rb

Elhambasir committed:
Create like.rb

Elhambasir committed:
Create comment.rb

Elhambasir committed:
Update comment.rb

Felix45 commented on app/models/user.rb:
```diff
@@ -0,0 +1,22 @@
+class User < ApplicationRecord
+  has_many :posts
+  has_many :comments
+  has_many :likes
+  has_many :liked_posts, through: :likes, source: :post
```
- [ Required ] Good job in setting up these associations, however, I think you missed one thing. Since you've named the `user_id` column in `posts`, `comments`, and `likes` models as `author_id` it would be better to also indicate the column that holds this association in the respective models. I would like to suggest that you refactor your code as shown below.

```ruby
 has_many :posts, foreign_key: 'author_id'
 has_many :comments, foreign_key: 'author_id'
 has_many :likes, foreign_key: 'author_id'

```

Elhambasir replied:
Ok thank you

Felix45 commented on app/models/user.rb:
```diff
@@ -0,0 +1,22 @@
+  validates :photo, presence: true
+  validates :bio, presence: true
+
+  User.create(name: 'Tom', photo: 'https://unsplash.com/photos/F_-0BxGuVvo', bio: 'Teacher from Mexico.')
+  User.create(name: 'Lilly', photo: 'https://unsplash.com/photos/F_-0BxGuVvo',
+              bio: 'Teacher from Poland.')
+
+  first_user = User.find_by(name: 'Tom')
+  first_user.name = 'Elham'
+  first_user.save
```
- [ Required ] I think this is not the best way to populate data in your application, kindly remove the lines of code I have highlighted from this file. In case you would like to populate data in your application I would suggest that you use the `seeds.rb` file in the `db` directory, this will make it easy for someone to set up your app using the `rails db:seed` command. 

## seed.rb
```ruby
  User.create(name: 'Tom', photo: 'https://unsplash.com/photos/F_-0BxGuVvo', 
                     bio: 'Teacher from Mexico.')
  User.create(name: 'Lilly', photo: 'https://unsplash.com/photos/F_-0BxGuVvo', 
                     bio: 'Teacher from Poland.')
```

Elhambasir replied:
Thank you for your clue and suggestion

Felix45 commented on app/models/post.rb:
```diff
@@ -0,0 +1,27 @@
+  has_many :likes
+  has_many :users, through: :likes
+
+  first_user = User.find_by(name: 'Tom')
+
+  Post.create(author: first_user, title: 'Hello', text: 'This is my first post')
+  Post.create(author: first_user, title: 'Hello', text: 'This is my second post')
+  Post.create(author: first_user, title: 'Hello', text: 'This is my third post')
+  Post.create(author: first_user, title: 'Hello', text: 'This is my fourth post')
+
+  Post.fine(1).delete
```
- [ Required ] I think this is not the best way to populate data in your application, kindly remove the lines of code I have highlighted from this file. In case you would like to populate data in your application I would suggest that you use the `seeds.rb` file in the `db` directory, this will make it easy for someone to set up your app using the `rails db:seed` command. 

## In seeds.rb
> In seeds.rb after creating the users add these lines of code
```ruby
  Post.create(author_id: 1, title: 'Hello', text: 'This is my first post')
  Post.create(author_id: 1, title: 'Hello', text: 'This is my second post')
  Post.create(author_id: 1, title: 'Hello', text: 'This is my third post')
  Post.create(author_id: 1, title: 'Hello', text: 'This is my fourth post')
```

Elhambasir replied:
Thank you

Felix45 commented on app/models/like.rb:
```diff
@@ -0,0 +1,10 @@
+class Like < ApplicationRecord
+  belongs_to :user
```
- [ Required ] I think since in your migrations the `user id` column in the `likes` table is named `author_id` you should specify that column when setting up this `belongs_to` association. I would like to suggest that you refactor this line of code as shown below.

```ruby
  belongs_to :author, class_name: 'User'
```

Elhambasir replied:
Ok

Felix45 commented on app/models/comment.rb:
```diff
@@ -0,0 +1,22 @@
+class Comment < ApplicationRecord
+  belongs_to :user
```
- [ Required ] I think since in your migrations the `user id` column in the `comments` table is named `author_id` you should specify that column when setting up this `belongs_to` association. I would like to suggest that you refactor this line of code as shown below.

```ruby
  belongs_to :author, class_name: 'User'
```

Elhambasir replied:
Got it

Felix45 commented on app/models/comment.rb:
```diff
@@ -0,0 +1,22 @@
+  Comment.create(post: first_post, user: first_user, text: 'Perfect')
+  Comment.create(post: first_post, user: first_user, text: 'I liked your post')
+  Comment.create(post: first_post, user: second_user, text: 'I love your post')
+  Comment.create(post: first_post, user: first_user, text: 'You look great!')
```
- [ Required ] I think this is not the best way to populate data in your application, kindly remove the lines of code I have highlighted. In case you would like to populate data in your application I would suggest that you use the `seeds.rb` file in the `db` directory, this will make it easy for someone to set up your app using the `rails db:seed` command. 

## In seeds.rb
> In seeds.rb after creating the posts add these lines of code
```ruby
  Comment.create(post_id: 1, author_id: 2, text: 'Hi Tom!')
  Comment.create(post_id: 1, author_id: 2, text: 'Your post is nice!')
```

Felix45 commented on app/models/user.rb:
```diff
@@ -0,0 +1,22 @@
+  def recent_posts(author_id)
+    author = User.find(author_id)
+    author.posts.order(created_at: :desc).limit(5)
+  end
```
- [ Required ] I think it would be better if you refactor this method for the following reasons, one, since the method is already in the `User` model and a `user` has many posts you don't have to pass the `author_id` to the `recent_posts` method because the association is already defined. Two, according to the project requirements the method is supposed to return only three recent posts by the user. I would like to suggest that you refactor this method as shown below.

## Example code
```ruby
def recent_post
  posts.order('created_at Desc').limit(3)
end
```

Felix45 commented on app/models/like.rb:
```diff
@@ -0,0 +1,10 @@
+  def self.update_like_counter(post_id)
+    post = Post.find(post_id)
+    post.update(like_counter: post.likes.count)
+  end
```
- [ Required ] I think it would be better if you refactor this method for the following reason since the method is already in the `Like` model and a `like` belongs to a `post` you don't have to pass the `post_id` to the `update_like_counter` method. 

You can refactor the method as suggested below, this example uses a [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - `after_save` to call the `#update_likes_counter` method every time a new like has been saved to database for a post.
## Example code
 ```ruby
class Like < ApplicationRecord

  belongs_to :author, class_name: 'User'
  belongs_to :post

  after_save :update_likes_counter
  
  def update_likes_counter
    post.increment!(:like_counter)
  end
end
```

Felix45 commented on app/models/comment.rb:
```diff
@@ -0,0 +1,22 @@
+  def self.update_comment_counter(post_id)
+    post = Post.find(post_id)
+    post.update(comment_counter: post.comments.count)
+  end
```
- [ Required ] I think it would be better if you refactor this method for the following reason since the method is already in the `Comment` model and a `comment` belongs to a `post` you don't have to pass the `post_id` to the `update_comment_counter` method because the association is already defined in the model. 

You can refactor the method as suggested below, this example uses a [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - `after_save` to call the `#update_comment_counter` method every time a new comment has been saved to the database for a post.
## Example code
 ```ruby
class Comment < ApplicationRecord
  belongs_to :author, class_name: 'User'
  belongs_to :post

  after_save :update_comments_counter
  
  def update_comments_counter
    post.increment!(:comment_counter)
  end
end
```

Felix45 commented on app/models/post.rb:
```diff
@@ -0,0 +1,27 @@
+  end
+
+  # A method which returns the 5 most recent comments for a given post.
+  def recent_comments(post_id)
+    post = Post.find(post_id)
+    post.comments.order(created_at: :desc).limit(5)
+  end
```
- [ Required ] I think it would be better if you refactor this method for the following reason, since the method is already in the `Post` model you don't have to pass the `post_id` to the `recent_comments` method. I would like to suggest that you refactor this method as shown below.

## Example code
```ruby
def recent_comments
  comments.order('created_at Desc').limit(5)
end
```

Felix45 commented on app/models/post.rb:
```diff
@@ -0,0 +1,27 @@
+  Post.fine(1).delete
+
+  # A method that updates the posts counter for a user.
+  def self.update_post_counter(user_id)
+    user = User.find(user_id)
+    user.update(post_counter: user.posts.count)
+  end
```
- [ Required ] I think it would be better if you refactor this method for the following reason since the method is already in the `Post` model and a `post` belongs to a `user` you don't have to pass the `user_id` to the `update_post_counter` method.

You can refactor the method as suggested below, this example uses a [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - `after_save` to call the `#update_post_counter` method every time a new post has been saved to the database.
## Example code
 ```ruby
class Post < ApplicationRecord
  belongs_to :author, class_name: 'User'
  has_many :comments
  has_many :likes

  after_save :update_post_counter
  
  def update_post_counter
    post.increment!(:post_counter)
  end
end
```

Felix45 commented on app/models/comment.rb:
```diff
@@ -0,0 +1,22 @@
+class Comment < ApplicationRecord
+  belongs_to :user
+  belongs_to :post, counter_cache: :comment_counter
```
- [ Required ] I think it would be better if you follow the project requirements when updating the `comment_counter` field I know that you can update the `comment_counter` field using `counter_cache` but according to the project requirements, this field should be updated using the `update_comments_counter` method so let's follow the project guidelines. Please see the comment I've left on the `update_comment_counter` method.

## Project requirements
![Screenshot from 2023-07-20 00-41-40](https://github.com/Elhambasir/blog-app/assets/16392046/81a9e4ad-3e3b-42ad-a3e2-5a39f7020bb8)

Felix45 commented on app/models/like.rb:
```diff
@@ -0,0 +1,10 @@
+class Like < ApplicationRecord
+  belongs_to :user
+  belongs_to :post, counter_cache: :like_counter
```
- [ Required ] I think it would be better if you follow the project requirements when updating the `like_counter` field I know that you can update the `like_counter` field using `counter_cache` but according to the project requirements, this field should be updated using the `update_like_counter` method so let's follow the project guidelines. Please see the comment I've left on the `update_like_counter` method.

## Project requirements
![Screenshot from 2023-07-20 00-40-21](https://github.com/Elhambasir/blog-app/assets/16392046/058b0084-81d0-4568-bf8c-92d028596571)

Felix45 commented on app/models/post.rb:
```diff
@@ -0,0 +1,27 @@
+class Post < ApplicationRecord
+  belongs_to :author, class_name: 'User', counter_cache: :post_counter
```
- [ Required ] I think it would be better if you follow the project requirements when updating the `post_counter` field, I know that you can update the `post_counter` field using `counter_cache` but according to the project requirements, this field should be updated using the `update_post_counter` method so let's follow the project guidelines. Please see the comment I've left on the `update_post_counter` method.

## Project requirements
![Screenshot from 2023-07-20 00-36-11](https://github.com/Elhambasir/blog-app/assets/16392046/2807df97-3496-4b61-a476-932131e188e6)

Felix45 reviewed (changes requested):
Hi @Elhambasir, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
<img src="https://camo.githubusercontent.com/59fa9d95c18dbd5a8e47490a9e4b48393f569685410f3c78d4d83a9764fa1392/68747470733a2f2f6d656469612e67697068792e636f6d2f6d656469612f74385950663341434b656f754c7a454976722f67697068792e676966" width="350px" />

## Highlights :heavy_check_mark: 
- [x] Git flow is used correctly :white_check_mark: 
- [x] Pull request has a descriptive summary and title :white_check_mark: 
- [x] Repository has a professional readme file :white_check_mark:   


## Required Changes ♻️
- [ Required ] I think it would be better if you made sure that all the migrations are running without an error, I encountered the error below when trying to run the migrations in your application. This means I could not use the rails console to interact with your App and it might also affect the functionality of your `App` in future milestones, I highly suggest that you fix this.

## Error when running migrations
![Screenshot from 2023-07-20 00-33-37](https://github.com/Elhambasir/blog-app/assets/16392046/7d30e560-d979-44ff-b754-54f9e2602a7c)




_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

Elhambasir committed:
Update comment.rb

Remove populated data from the application

Elhambasir committed:
Update post.rb

Remove populated data from the application
Use method to update the counter instead of counter_cache

Elhambasir committed:
Update user.rb

Remove populated data from the application.
Indicate the column that holds this association in the respective models

Elhambasir committed:
Create schema.rb

Elhambasir committed:
Update seeds.rb

Elhambasir committed:
Update 20230718142009_add_user_ref_to_comments.rb

Elhambasir committed:
Update like.rb

Use method to update counter instead of counter_cache

Elhambasir committed:
Merge branch 'dev' into processing-data-in-models

danmainah commented on app/models/like.rb:
```diff
@@ -0,0 +1,11 @@
+class Like < ApplicationRecord
+  belongs_to :user, class_name: 'User'
```
- [ ] @Elhambasir Greta job implementing the changes requested, however in Like model you have stated that it belongs to the user instead of author, to be inline with your schema, it produces error below when I try to create a like
![image](https://github.com/Elhambasir/blog-app/assets/61408860/ad355a63-6b6c-4502-85d6-d897c7a6acdd)

Elhambasir replied:
Fixed

danmainah commented on db/seeds.rb:
```diff
@@ -5,3 +5,28 @@
 #
 #   movies = Movie.create([{ name: "Star Wars" }, { name: "Lord of the Rings" }])
 #   Character.create(name: "Luke", movie: movies.first)
+
+Comment.create(post: first_post, author: second_user, text: 'Hi Tom!')
+Comment.create(post: first_post, author: second_user, text: 'Your post is nice!')
+Comment.create(post: first_post, author: first_user, text: 'Perfect')
+Comment.create(post: first_post, author: first_user, text: 'I liked your post')
+Comment.create(post: first_post, author: second_user, text: 'I love your post')
+Comment.create(post: first_post, author: first_user, text: 'You look great!')
+
+
+first_user = User.find(1)
+second_user = User.find(2)
+
+first_post = Post.create(author: first_user, title: 'Hello', text: 'This is my first post')
+second_post = Post.create(author: first_user, title: 'Hello', text: 'This is my second post')
+third_post = Post.create(author: first_user, title: 'Hello', text: 'This is my third post')
+fourth_post = Post.create(author: first_user, title: 'Hello', text: 'This is my fourth post')
+
+first_post = Post.find(3)
+second_post = Post.find(4)
+third_post = Post.find(5)
+fourth_post = Post.find(6)
+
+User.create(name: 'Tom', photo: 'https://unsplash.com/photos/F_-0BxGuVvo', bio: 'Teacher from Mexico.')
+User.create(name: 'Lilly', photo: 'https://unsplash.com/photos/F_-0BxGuVvo',
+            bio: 'Teacher from Poland.')
```
- [x] @Elhambasir Greate work implementing the changes requested, however for it to be effective especially when you create a new database, kindly consider rearranging the code, the code to create users should be at the top, followed by code to create posts and finally the comments code.

Elhambasir replied:
Ok thank you

danmainah reviewed (changes requested):
Hi @Elhambasir , 

Great work on making the changes requested by a previous reviewer 👏  👏 👏
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
✔️  Implemented changes requested to ensure the user can run migrations successfully
✔️ Added all methods requested correctly.

 # Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

Elhambasir committed:
Change user to author Arrange the data inside seed.rb

Elhambasir committed:
Merge branch 'processing-data-in-models' of https://github.com/Elhambasir/blog-app into processing-data-in-models

mikethreels reviewed (approved):
Hi Elham, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
- Good job implementing the requested changes.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
